File: hand.py
"""
Representation for Cribbage-specific hand.
"""
import itertools
import functools
from cards import Card

class Hand:
    """
    Represents a hand of cards.
    """
    def __init__(self, hand_cards, cut_card=None, crib=False):
        if cut_card:
            self.cards = hand_cards + [cut_card]
        else:
            self.cards = hand_cards
        for card in self.cards:
            if not isinstance(card, Card):
                raise RuntimeError('Illegal value ("' + str(card) + '")')
        self.hand_cards = hand_cards
        self.cut_card = cut_card
        self.crib = crib

    def __str__(self):
        string = 'Hand: [' + str(self.hand_cards[0])
        for card in self.hand_cards[1:]:
            string += ', ' + str(card)
        if self.cut_card:
            string += ' | ' + str(self.cut_card)
        if self.crib:
            string += ' | crib'
        string += ']'
        return string

    def count(self):
        """
        Returns the point count for the hand.
        """
        points = 0

        # all possible subsets of cards within the hand
        subsets = functools.reduce(
            lambda x, y: x + y,
            [list(itertools.combinations(self.cards, x + 1)) for x in range(len(self.cards))]
        )

        for subset in subsets:
            subset_ranks = list(map(lambda card: card.rank, list(subset)))

            # count fifteens
            if sum(list(map(lambda card: card.value, list(subset)))) == 15:
                print('  fifteen')
                points += 2

            # count pairs
            if len(subset) == 2 and subset_ranks[0] == subset_ranks[1]:
                print('  pair')
                points += 2

            # count runs
            smallest_rank = min(subset_ranks)
            run_length = 1
            while smallest_rank + run_length in subset_ranks:
                run_length += 1
            card_ranks = list(map(lambda x: x.rank, self.cards))
            if run_length >= 3 and run_length == len(subset) and smallest_rank + run_length not in card_ranks and smallest_rank - 1 not in card_ranks:
                print('  run of ' + str(run_length))
                points += run_length

        # count flushes
        flush_suit = functools.reduce(lambda x, y: x if x == y else False, map(lambda x: x.suit, self.hand_cards))
        cut_matches = self.cut_card.suit == flush_suit
        flush_points = 0
        if cut_matches:
            flush_points = len(self.cards)
        elif not self.crib and flush_suit:
            flush_points = len(self.cards) - 1
        if flush_points > 0:
            print('  ' + str(flush_points) + '-card flush')
            points += flush_points

        # count knob
        if len([card for card in self.hand_cards if card.rank == 11 and card.suit == self.cut_card.suit]) > 0:
            print('  knob')
            points += 1

        return points

File: src/server/game/hand.ts
/**
 * @brief Representation for Cribbage-specific hand.
 */
import { Card, Suit } from './cards';
import { PowerSet } from 'js-combinatorics';
import { Util } from './util';

/**
 * @brief Represents a hand of cards.
 */
export class Hand {
    private readonly cards: Card[];

    public constructor(
        private readonly handCards: Card[],
        private readonly cutCard: Card | null = null,
        private readonly crib: boolean = false,
    ) {
        if (cutCard) {
            this.cards = [...handCards, cutCard];
        } else {
            this.cards = [...handCards];
        }
        this.cards.forEach((card) => {
            if (!(card instanceof Card)) {
                throw new Error(`Illegal value ("${card}")`);
            }
        });
    }

    /**
     * @brief Returns the point count for the hand.
     */
    public readonly count = (verbose = false) => {
        let points = 0;

        // all possible subsets of cards within the hand
        const subsets: Card[][] = [...new PowerSet(this.cards)];

        subsets.forEach((subset) => {
            const subsetRanks = subset.map((card) => {
                return card.rank;
            });
            const subsetValues = subset.map((card) => {
                return card.value;
            });
            const cardRanks = this.cards.map((card) => {
                return card.rank;
            });

            // count fifteens
            if (
                subsetValues.reduce((total, value) => {
                    return total + value;
                }, 0) === 15
            ) {
                Util.log(verbose, 'fifteen');
                points += 2;
            }

            // count pairs
            if (subset.length === 2 && subsetRanks[0] === subsetRanks[1]) {
                Util.log(verbose, 'pair');
                points += 2;
            }

            // count runs
            const smallestRank = Math.min(...subsetRanks);
            const checkRun = (runLength: number) => {
                return subsetRanks.some((rank) => {
                    return rank === smallestRank + runLength;
                });
            };
            let runLength = 1;
            while (checkRun(runLength)) {
                runLength += 1;
            }
            if (
                runLength >= 3 &&
                runLength === subset.length &&
                cardRanks.every((rank) => {
                    return (
                        rank !== smallestRank + runLength &&
                        rank !== smallestRank - 1
                    );
                })
            ) {
                Util.log(verbose, `run of ${runLength}`);
                points += runLength;
            }
        });

        // count flushes
        const flushSuit = Suit.SUITS.find((suit) => {
            return this.handCards.every((card) => {
                return card.suit === suit;
            });
        });
        const cutMatches = this.cutCard?.suit === flushSuit;
        const flushPoints = cutMatches
            ? this.cards.length
            : !this.crib && flushSuit
            ? this.cards.length - 1
            : 0;
        if (flushPoints > 0) {
            Util.log(verbose, `${flushPoints}-card flush`);
            points += flushPoints;
        }

        // count nob
        if (
            this.handCards.some((card) => {
                return card.rank === 11 && card.suit === this.cutCard?.suit;
            })
        ) {
            Util.log(verbose, 'nob');
            points += 1;
        }

        return points;
    };

    public readonly toString = () => {
        return `Hand: [${this.handCards.join(', ')}${
            this.cutCard ? ` | ${this.cutCard}` : ''
        }${this.crib ? ' | crib' : ''}]`;
    };
}
